Provide fallbacks for command line

gassmoeller opened the pull request:
A set of patches that should work around some issues that occur, if `system()` (to do something on the command line) does not work for some reason. This is the case for compute nodes on IBM's BlueGene/Q architecture.

The inclusion of the different namespace in Utilities is only coarsely related. It would be nice, if I would not need to rename every `Utilities::something` function into `dealii::Utilities::something` just because I include the `aspect::Utilities` header file. Is this solution ok?

bangerth commented:
I think the solution to import `dealii::Utilities` into `aspect::Utilities` is acceptable to me. @tjhei ?

bangerth commented on source/postprocess/visualization.cc:
```diff
@@ -531,46 +531,49 @@ namespace aspect
-            tmp_filename = *filename;
-          }
-      }
+      if (system(0) != 0)
```
So is this is how this works on the BG/Q? The `system` function is declared in header files, but it will always yield an error?

I find the call `system(0)` odd to understand. Maybe write it as `system((char*)0)`?

tjhei replied:
yes, a comment here would be nice.

bangerth commented:
The rest looks good. Can you rewrite the one call to `system` and then merge?

tjhei commented:
> I think the solution to import dealii::Utilities into aspect::Utilities is acceptable to me. @tjhei ?

Awkward. I think dealii::Utilities has different content than aspect:Utilities. I don't mind spelling it out as dealii::Utilities, but understand that not being consistent is bad. I don't know of a better solution though.

tjhei commented on include/aspect/utilities.h:
```diff
@@ -44,7 +44,7 @@ namespace aspect
   namespace Utilities
   {
-    using namespace dealii;
+    using namespace dealii::Utilities;
```
Is there a reason you are removing the "using namespace dealii" line? This should stay here I think.

bangerth commented:
@gassmoeller -- what's the current status of this?

gassmoeller commented:
I need to find some time to check a few things on the BlueGene. I will let you know when its finished (should be soon).

bangerth commented:
I think I'd like to merge this before 1.4. @gassmoeller -- when you have a second, can you rebase, add the comment, and see whether it all works?

gassmoeller commented:
I have checked the remaining things, it now works. Also added the comment and added an entry to changes.h. When the tester is happy this should be ready to merge.

bangerth commented:
I think you'll need to run the indenter.

bangerth commented on source/simulator/checkpoint_restart.cc:
```diff
@@ -39,7 +40,17 @@ namespace aspect
+          error = rename(old_name.c_str(),new_name.c_str());
+        }
 
       AssertThrow (error == 0, ExcMessage(std::string ("Can't move files: ")
```
Move this assert directly after the place where you set `error`, and replicate it also in the `if (Utilities::fexists(new_name))` clause -- currently, if you get in there, you call `remove`, set `error`, but then immediately overwrite it. You should check it after both the `remove` and the `rename` call.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -764,12 +764,19 @@ namespace aspect
+        // mkdir is not available on the command-line, fall back to mkdir().
+        if (error != 0)
+          {
+            error = mkdir(output_directory.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
```
The difficulty that prompted me to use the `system("mkdir -p")` call is that the `mkdir` function can only create one level of subdirectories, if I recall this correctly. On the other hand, `mkdir -p` creates as many levels as necessary.

The distinction is relevant in cases where someone sets the output directory to `./output/run-13/test-42`, but neither `output/` doesn't exist yet. This is slightly complicated with just `std::mkdir` -- a solution is here, for example: http://stackoverflow.com/questions/675039/how-can-i-create-directory-tree-in-c-linux . (It would be nice if we could use the boost::filesystem function, but that's not part of our piece of Boost :-( . There is a nice algorithm further down that page, posted by user "Patrick", though.)

gassmoeller replied:
When I implement that functionality is there any reason to keep the old version that uses system()? Removing the old version would save us this lengthy comment. Also would you prefer the new function mkpath to be in an anonymous namespace of parameters.cc, or should I put it in utilities.cc?

bangerth replied:
Yes, good point -- if we had such a function, we won't need the old code any more.

I think calling it `mkdirp` and putting it into namespace `Utilities` actually makes sense. Thanks for working on this!

gassmoeller commented:
This new function should do the trick. If the tester is happy, feel free to take another look. I checked for most cases on my system and the mkdirp seems to work fine, it is hard to put this in a testcase though, because it does not fit in our usual test folder structure. What is your opinion on the default permissions of the output folder? For now I set it to 755.

bangerth commented:
Great, thanks for sticking with me through the many iterations here!

bangerth commented on source/simulator/checkpoint_restart.cc:
```diff
@@ -39,11 +40,25 @@ namespace aspect
+                                                               + ", although it seems to exist.")));
+            }
+
+          error = rename(old_name.c_str(),new_name.c_str());
```
Ah, crap, I recognize another problem with this patch now. Previously, we used `system("mv ...")`. The `mv` command renames or moves files, between directories and if necessary from one file system to another. But the `rename` command only does the former. Among its error codes it says
```
      EXDEV  oldpath  and  newpath  are  not on the same mounted file system.  (Linux permits a file system to be mounted at multiple points, but rename() does not
              work across different mount points, even if the same file system is mounted on both.)
```
In other words, `rename` will not be able to move a file from `/tmp` to a subdirectory of `HOME` on many systems. What do we do?

tjhei replied:
but ``move_file`` is only used to rename checkpoint .old files in the same directory, right?

gassmoeller replied:
Correct. I do not think this should create problems.

bangerth replied:
Don't we also use the function when moving `.vtu` files that we temporarily write to tmp storage and then move to their permanent place? I may be completely off here...

bangerth replied:
I am completely off here. Forget about it.

gassmoeller replied:
I think we decided against using it for temporary visualization files for exactly the reason you mentioned (I think I wanted to use it and you pointed out the problem :smile:, so you saved yourself some trouble here). Currently all places should first try the `system()` command, and then have a fallback if that does not work.

bangerth replied:
I need to stop speculating about what may or may not be in the code at any given moment. There seems to no longer be a correlation between what I think is in there and what is actually in there...
